### ISSUE TITLE
[PROPOSAL] Embed down scripts inside the database

### ISSUE THREAD

**Issue by nicolabovolato** [golang-migrate/migrate#923]
In my opinion, this is a feature that focuses on a very specific problem, yet it can fix it without manual intervention, which right now is what's needed.

**Is your feature request related to a problem? Please describe.**

When a older version of a service (that includes less migrations than the current one) is deployed, the down steps will always fail (if the migrations are put alongside the code). This is because migrations are only stored outside of the database and so the required files for the down steps are missing.

Let's make an example:

```
📁
├─ 📁 migrations
├─ 📄 go.mod
├─ 📄 go.sum
└─ 📄 main.go
```

There's a go service. The migrations are put alongside the code in the same repo.

Inside main.go, the Up migrations are ran every time the service starts

```go
//...
func main() {

	migrator, err := migrate.New(MIGRATIONS_URL, DB_URL)
	if err != nil {
		panic(err)
	}

	err = migrator.Up()
	if err != migrate.ErrNoChange {
		panic(err)
	}

    //...
}
```

The repo follows github flow: for every new feature a PR is opened to master.

A service like Heroku is connected to the repo. Every time there's a new commit on master the updated service is deployed on prod.

First commit, these migrations are present in master. The service is deployed.

```
📁 migrations
├─ 📄 000001_init.down.sql
└─ 📄 000001_init.up.sql
```

A new feature that involves migrations is implemented, the PR is approved and it gets merged into master. The service is deployed.

```
📁 migrations
├─ 📄 000002_new_feature.down.sql
└─ 📄 000002_new_feature.up.sql
```

Some time pass and the feature is not relevant anymore. No other changes have been published onto master, so the latest merge commit just gets reverted.

The service deploy fails in prod with the following error:

```
panic: no migration found for version 2: read down for version 2 .: file does not exist
```

What happened?

- Migrator finds the latest version in our migrations source, which is version 1.
- Because the database schema_migrations table has version 2 it tries to do a down from version 2 to version 1.
- Version 2's down script is not present in our migrations source so the command fails.

Manual intervention is required on the database to fix the error.

**Describe the solution you'd like**

I think a nice addition to migrate would be to store the down migrations inside the database in a separate table/collection.
If a down migration is not found on the source path, it will be retrieved from the db and consequently applied.

Having this functionality means no manual intervention on the db when downgrading a service.
Obviously, if a down migration was written incorrectly it will still fail.

**Describe alternatives you've considered**

I thought about storing migrations in another repo or on S3.
That is not a quick fix, but a complete paradigm shift in my opinion.

Having migrations alongside your code allows you to have a complete overview of a service's history and current state. It is also quicker to make changes.
I think separating them it's just not worth it, that even more if you work with microservices and need to manage 20 of them.

You can still have the migrations alongside your code and point the source env var to the repo, but that would having an env var point to a specific commit.
And that would make managing env vars (and the ci) something way more complex than it should be.

**Comment by obalunenko** [comment]
@nicolabovolato I think that you can try to decompose feature development into two steps:
- PR with migrations
- PR with code changes (only after PR with migrations approved and merged)
Migrations should be written in backward-compatible non-breaking way.
In this case when you want to revert some changes - it will be always possible to do in easy way.

**Comment by nicolabovolato** [comment]
@obalunenko What do you mean with the first point? I can't understand

**Comment by obalunenko** [comment]
@nicolabovolato I mean that database migrations should be delivered separately to the master branch.

**Comment by nicolabovolato** [comment]
@obalunenko Thanks, now I get it. I was thinking of a completely different context.
I guess your solution has a valid point. I think it's also a far better approach than actually implementing my proposal.

Also, another approach, that fits well with your solution, might be to not have a down migration but instead write its contents in the next up migration. This way you cannot end up with the error described above.

I'll close the issue!
